packages/edit-site/src/components/sync-state-with-url/use-sync-canvas-mode-with-url.js

```diff
@@ -20,22 +20,52 @@ export default function useSyncCanvasModeWithURL() {
 	);
 	const { setCanvasMode } = unlock( useDispatch( editSiteStore ) );
 	const currentCanvasMode = useRef( canvasMode );
-	const { canvas: canvasInUrl = 'view' } = params;
+	const { canvas: canvasInUrl } = params;
 	const currentCanvasInUrl = useRef( canvasInUrl );
+	const currentUrlParams = useRef( params );
+	useEffect( () => {
+		currentUrlParams.current = params;
+	}, [ params ] );
+
 	useEffect( () => {
 		currentCanvasMode.current = canvasMode;
-		if ( currentCanvasMode !== currentCanvasInUrl ) {
+		if ( canvasMode === 'init' ) {
+			return;
+		}
+
+		if (
+			canvasMode === 'edit' &&
+			currentCanvasInUrl.current !== canvasMode
+		) {
+			history.push( {
+				...currentUrlParams.current,
+				canvas: 'edit',
+			} );
+		}
+
+		if (
+			canvasMode === 'view' &&
+			currentCanvasInUrl.current !== undefined
+		) {
 			history.push( {
-				...params,
-				canvas: canvasMode,
+				...currentUrlParams.current,
+				canvas: undefined,
```

Comment:
I'm still debugging but do you remember why we needed to set `undefined` here? There is a problem when we want to push to history something like:
```
history.push( {
	postId: post.id,
	postType: post.type,
	canvas: 'edit',
} );
```
The above example is used in pages list(click the title or `edit` in actions) and creates unexpected results, since at first the canvas mode is `view` in the state and we update it here. So an example like the above doesn't seem to have a good way of being achieved. and adds one more entry to history too. 

I mean, if we want to push to history `edit` canvas, should we just use `setCanvasMode( 'edit' )` and leave the `canvas` param get handled by this hook? 🤔

Reply:
When I first wrote this PR, unsetting the canvas from the url meant you're going to "view" mode. I think that's good because on `site-editor.php` you don't specify anything and it loads and there's no rewriting of the url.  IMO, `view` should never be present in the URL, it's either not set (undefined) or "edit".

Reply:
I don't disagree with the `view` being absent from url. The issue still exists though when we want to go directly to `edit` from `view` and it seems we probably can't avoid using `setCanvasMode` in such cases. I'll explore more and open a PR.

Reply:
I got that, that's what I'm trying to figure out how to solve in the best possible way. Thanks!